### PR TITLE
SslBump: Support parsing GREASEd (and future) TLS handshakes

### PR DESCRIPTION
**Pull request by chtsanti**
A peeking or staring Squid aborted TLS connections containing a GREASE
version in the supported_versions handshake extension (e.g., 0x3a3a).
Here is a sample cache.log error (debug_options ALL,1 83,2):

    parseTlsHandshake: ... check failed: vMajor == 3
        exception location: Handshake.cc(119) ParseProtocolVersion

The same problem would apply to some other "unsupported" (and currently
non-existent) TLS versions (e.g., 0x0400). The growing popularity of
GREASE values exposed the bug (just like RFC 8710 was designed to do).

Squid now ignores all unsupported-by-Squid TLS versions in handshake
extensions. Squid still requires a supported TLS version for
framing-related handshake fields -- no changes there.

It is difficult to define "supported" in this context precisely because
a peeking Squid only observes the TLS handshake. Our handshake parser
may report the following versions: SSL v2.0, SSL v3.0, and TLS v1.x
(with x >= 0). This logic allows us to safely parse the handshake (no
framing errors) and continue to make version-based decisions like
disabling OpenSSL TLS v1.3 support when the agents are negotiating TLS
v1.3+ (master cd29a42). Also, access logs benefit from this "support"
going beyond TLS v1.2 even though Squid cannot bump most TLS v1.3+
handshakes.

Squid was and still is compliant with the "MUST NOT treat GREASE values
differently from any unknown value" requirement of RFC 8710.

Also added source code comments to mark other places where unsupported
(for some definition of "support") TLS values, including GREASE values
may appear.

FWIW, we have implemented but then eventually removed GREASE value
detection code. Folks studying debugging logs (where GREASE values may
appear) will probably either recognize their easy-to-remember 0x?A?A
pattern or not really know/care about RFC 8710 and its special values.

This is a Measurement Factory project.

### CODE REVIEW

**Comment by chtsanti** [comment]
RFC 8701 says
> When processing a ClientHello, servers MUST NOT treat GREASE values
> differently from any unknown value.

This is not possible for supported protocols extension. Because we need to log the maximum supported version offered by client (important info, even if the maximum supported version is not yet supported by squid or not exist/implemented/defined yet) but not GREASEd versions because they are exist in any HELLO client message and will overwrite the important info.

File: src/ssl/bio.cc [hunk 1]
```diff
@@ -714,6 +714,9 @@ applyTlsDetailsToSSL(SSL *ssl, Security::TlsDetails::Pointer const &details, Ssl
                 if (!strCiphers.isEmpty())
                     strCiphers.append(":");
                 strCiphers.append(SSL_CIPHER_get_name(c));
+            } else {
+                debugs(83, 7, "ignoring " << (Security::Greased(cipherId) ? "GREASEd" : "unsupported") <<
+                       " cipher: " << cipherId);
```
Comment (rousskov) [inline src/ssl/bio.cc:719]: @chtsanti, do we need to keep adding unsupported and GREASEd cipher IDs to `details->ciphers`? This code does not seem to need them. I know adjustSSL() checks ciphers, but there we could just remember whether we saw any unsupported ones during parsing -- we already do the same for extensions with `details->unsupportedExtensions`.

Comment (chtsanti) [reply src/ssl/bio.cc:719]: What if in the future implement ciphers based acls or an access list which allow/deny special ciphers? In these cases maybe it is required to tke in account GREASEd values.

Comment (rousskov) [reply src/ssl/bio.cc:719]: Thank you for a prompt answer. I agree that such a future is possible, perhaps even likely. I will be ableto move forward with that information in mind.

FWIW, I do not like the inconsistency between our approaches to dealing with unsupported and GREASEd cipher IDs and dealing with unsupported and GREASEd TLS versions. However, resolving that inconsistency is probably outside this issue scope. I do not plan on doing that here.

**Comment by rousskov** [comment]
This became official PR 663.

**Pull request closed without merging**